File: libsrc/meshio/Xdmf.cc
// -*- C++ -*-
//
// ======================================================================
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University of Chicago
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010 University of California, Davis
//
// See COPYING for license information.
//
// ======================================================================
//

#include "Xdmf.hh" // implementation of class methods

#include "pylith/utils/array.hh" // USES int_array, string_vector

#include <string> // USES std::string
#include <stdexcept> // USES std::runtime_error
#include <iostream> // USES std::cout
#include <sstream> // USES std::ostringstream
#include <cassert> // USES assert()

// ----------------------------------------------------------------------
// Default constructor.
pylith::meshio::Xdmf::Xdmf(void)
{ // constructor
} // constructor

// ----------------------------------------------------------------------
// Destructor
pylith::meshio::Xdmf::~Xdmf(void)
{ // destructor
} // destructor

// ----------------------------------------------------------------------
// Write Xdmf file associated with HDF5 file.
void
pylith::meshio::Xdmf::write(const char* filenameXdfm,
			    const char* filenameHDF5)
{ // write
  std::string cellType = 0;
  int numCells = 0;
  int numCorners = 0;
  int numVertices = 0;
  int spaceDim = 0;
  int numTimeSteps = 0;
  double* timeStamps = 0;
  int numFields = 0;
  string_vector fieldName;
  string_vector fieldVectorFieldType;
  string_vector fieldType;
  int_array fieldNumPoints;
  int_array fieldFiberDim;

  if (spaceDim == 1) {
    std::cout
      << "WARNING: Xdmf grids not defined for 1-D domains.\n"
      << "Skipping creation of Xdmf file associated with HDF5 file '"
      << filenameHDF5 << "'" << std::endl;
    return;
  } // if

  _file
    << "<?xml version=\"1.0\" ?>\n"
    << "<!DOCTYPE Xdmf SYSTEM \"Xdmf.dtd\" [\n"
    << "<!ENTITY HeavyData \"" << filenameHDF5 << "\">\n"
    << "]>\n"
    << "\n"
    << "<Xdmf>\n"
    << "  <Domain Name=\"domain\">\n";

  _writeDomainCells(numCells, numCorners);
  _writeDomainVertices(numVertices, spaceDim);

  _file
    << "<!-- ============================================================ -->\n";

  _writeTimeStamps(timeStamps, numTimeSteps);

  _file
    << "<Grid Name=\"domain\" GridType=\"Uniform\">\n";
  for (int iTimeStep=0; iTimeStep < numTimeSteps; ++iTimeStep) {
    _writeGridTopology(cellType.c_str(), numCells);
    _writeGridGeometry(spaceDim);
    for (int iField=0; iField < numFields; ++iField) {
      if (2 == spaceDim && 
	  std::string("vector") == fieldVectorFieldType[iField]) {
	for (int component=0; component < spaceDim; ++component)
	  _writeGridAttributeComponent(fieldName[iField].c_str(),
				       fieldVectorFieldType[iField].c_str(),
				       fieldType[iField].c_str(),
				       numTimeSteps, fieldNumPoints[iField],
				       fieldFiberDim[iField],
				       iTimeStep, component);
      } else {
	_writeGridAttribute(fieldName[iField].c_str(),
			    fieldVectorFieldType[iField].c_str(),
			    fieldType[iField].c_str(), numTimeSteps, 
			    fieldNumPoints[iField], fieldFiberDim[iField],
			    iTimeStep);
      } // if/else
    } // for
  } // for

  _file
    << "      </Grid>\n"
    << "    </Grid>\n"
    << "  </Domain>\n"
    << "</Xdmf>\n";
  _file.close();
} // write

// ----------------------------------------------------------------------
// Write domain cell information.
void
pylith::meshio::Xdmf::_writeDomainCells(const int numCells,
					const int numCorners)
{ // _writeDomainCells
  _file
    << "    <DataItem Name=\"cells\"\n"
    << "	      ItemType=\"Uniform\"\n"
    << "	      Format=\"HDF\"\n" 
    << "	      NumberType=\"Float\" Precision=\"8\"\n"
    << "	      Dimensions=\"" << numCells << " " << numCorners << "\">\n"
    << "      &HeavyData;:/topology/cells\n"
    << "    </DataItem>\n";

} // _writeDomainCells

// ----------------------------------------------------------------------
// Write domain vertices information.
void
pylith::meshio::Xdmf::_writeDomainVertices(const int numVertices,
					   const int spaceDim)
{ // _writeDomainVertices
  _file
    << "    <DataItem Name=\"vertices\"\n"
    << "	      Format=\"HDF\"\n"
    << "	      Dimensions=\"" << numVertices << " " << spaceDim << "\">\n"
    << "      &HeavyData;:/geometry/vertices\n"
    << "    </DataItem>\n";
} // _writeDomainVertices

// ----------------------------------------------------------------------
// Write time stamps.
void
pylith::meshio::Xdmf::_writeTimeStamps(const double* timeStamps,
				       const int numTimeSteps)
{ // _writeTimeStamps
  if (numTimeSteps > 0) {
    assert(timeStamps);
  } // if
} // _writeTimeStamps

// ----------------------------------------------------------------------
// Write grid topology information.
void
pylith::meshio::Xdmf::_writeGridTopology(const char* cellType,
					 const int numCells)
{ // _writeGridTopology
  _file
    << "	<Topology\n"
    << "	   TopologyType=\"Triangle\"\n"
    << "	   NumberOfElements=\"" << numCells << "\">\n"
    << "	  <DataItem Reference=\"XML\">\n"
    << "	    /Xdmf/Domain/DataItem[@Name=\"cells\"]\n"
    << "	  </DataItem>\n"
    << "	</Topology>\n";
} // _writeGridTopology

// ----------------------------------------------------------------------
// Write Grid geometry.
void
pylith::meshio::Xdmf::_writeGridGeometry(const int spaceDim)
{ // _writeGridGeometry
  assert(2 == spaceDim || 3 == spaceDim);

  const char* geomType = (spaceDim == 3) ? "XYZ" : "XY";

  _file
    << "	<Geometry GeometryType=\"" << geomType << "\">\n"
    << "	  <DataItem Reference=\"XML\">\n"
    << "	    /Xdmf/Domain/DataItem[@Name=\"vertices\"]\n"
    << "	  </DataItem>\n"
    << "	</Geometry>\n";
} // _writeGridGeometry

// ----------------------------------------------------------------------
// Write grid attribute.
void
pylith::meshio::Xdmf::_writeGridAttributeComponent(const char* name,
						   const char* vectorFieldType,
						   const char* center,
						   const int numTimeSteps,
						   const int numPoints,
						   const int fiberDim,
						   const int iTime,
						   const int component)
{ // _writeGridAttribute
  std::string componentName = "unknown";
  switch (component) {
  case 0:
    componentName = std::string("x-") + std::string(name);
    break;
  case 1:
    componentName = std::string("y-") + std::string(name);
    break;
  case 2:
    componentName = std::string("z-") + std::string(name);
    break;
  default:
    { // default
      std::ostringstream msg;
      msg << "Unknown component " << component << " while writing Xdmf file.";
      std::cerr << msg.str() << std::endl;
      assert(0);
    throw std::logic_error(msg.str());
    } // default
  } // switch
  _file
    << "	<Attribute\n"
    << "	   Name=\"" << componentName << "\"\n"
    << "	   Type=\""<< vectorFieldType << "\"\n"
    << "	   Center=\"" << center << "\">\n"
    << "          <DataItem ItemType=\"HyperSlab\"\n"
    << "		    Dimensions=\"1 " << numPoints << " 1\"\n"
    << "		    Type=\"HyperSlab\">\n"
    << "            <DataItem\n"
    << "	       Dimensions=\"3 3\"\n"
    << "	       Format=\"XML\">\n"
    << "              0 0 "<< component << "\n"
    << "              1 1 1\n"
    << "              1 " << numPoints << " 1\n"
    << "	    </DataItem>\n"
    << "	    <DataItem\n"
    << "	       DataType=\"Float\" Precision=\"8\"\n"
    << "	       Dimensions=\""
    << numTimeSteps << " " << numPoints << " " << fiberDim << "\"\n"
    << "	       Format=\"HDF\">\n"
    << "	      &HeavyData;:/vertex_fields/" << name << "\n"
    << "	    </DataItem>\n"
    << "	  </DataItem>\n"
    << "	</Attribute>\n";
} // _writeGridAttribute

// ----------------------------------------------------------------------
// Write grid attribute.
void
pylith::meshio::Xdmf::_writeGridAttribute(const char* name,
					  const char* vectorFieldType,
					  const char* center,
					  const int numTimeSteps,
					  const int numPoints,
					  const int fiberDim,
					  const int iTime)
{ // _writeGridAttribute
  _file
    << "	<Attribute\n"
    << "	   Name=\"" << name << "\"\n"
    << "	   Type=\"" << vectorFieldType << "\"\n"
    << "	   Center=\"" << center << "\">\n"
    << "          <DataItem ItemType=\"HyperSlab\"\n"
    << "		    Dimensions=\"1 " << numPoints << " 1\"\n"
    << "		    Type=\"HyperSlab\">\n"
    << "            <DataItem\n"
    << "	       Dimensions=\"3 3\"\n"
    << "	       Format=\"XML\">\n"
    << "              0 0 0\n"
    << "              1 1 1\n"
    << "              1 " << numPoints << " 1\n"
    << "	    </DataItem>\n"
    << "	    <DataItem\n"
    << "	       DataType=\"Float\" Precision=\"8\"\n"
    << "	       Dimensions=\""
    << numTimeSteps << " " << numPoints << " " << fiberDim << "\"\n"
    << "	       Format=\"HDF\">\n"
    << "	      &HeavyData;:/vertex_fields/" << name << "\n"
    << "	    </DataItem>\n"
    << "	  </DataItem>\n"
    << "	</Attribute>\n";
} // _writeGridAttribute


// End of file

File: pylith/meshio/Xdmf.py
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University at Buffalo
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2021 University of California, Davis
#
# See LICENSE.md for license information.
#
# ----------------------------------------------------------------------
#

# @file pythia.pyre/meshio/Xdmf.py
##
# @brief Python class for Xdmf metadata file associated with an HDF5 file.
##
# Factory: xdmf


class Field(object):
    """Python object for data associated with vertex or cell field in HDF5 file.
    """
    DOMAIN_VERTICES = "Node"
    DOMAIN_CELLS = "Cell"

    groupToDomain = {
        "vertex_fields": DOMAIN_VERTICES,
        "cell_fields": DOMAIN_CELLS,
    }
    domainToGroup = dict((v, k) for k, v in groupToDomain.items())

    def __init__(self):
        self.name = None
        self.vectorFieldType = None
        self.data = None
        self.domain = None
        return


# Xdmf class
class Xdmf(object):
    """Python class for Xdmf metadata file associated with an HDF5 file.
    """

    # PUBLIC METHODS /////////////////////////////////////////////////////

    def __init__(self):
        """Constructor.
        """
        self.file = None
        return

    def write(self, filenameH5, filenameXdmf=None, verbose=False):
        """Write Xdmf file corresponding to given HDF5 file.
        """
        if not filenameXdmf:
            filenameXdmf = filenameH5.replace(".h5", ".xmf")
        if verbose:
            print("Generating %s..." % filenameXdmf)

        import h5py
        import os
        if not os.path.isfile(filenameH5):
            raise IOError(
                "Cannot create Xdmf file for HDF5 file '%s'. File not found." % filenameH5)
        self.h5 = h5py.File(filenameH5, "r")

        if self._getSpaceDim() == 1:
            print("WARNING: Xdmf grids are not defined for 1-D domains.\n"
                  "Skipping creation of Xdmf file for HDF5 file '%s'." % filenameH5)
            self.h5.close()
            return

        self.file = open(filenameXdmf, "w")

        # Header
        self._openXdmf()

        # Domain
        cells = self.h5["/topology/cells"]
        vertices = self.h5["/geometry/vertices"]
        self._openDomain(cells, vertices)

        timeStamps = self._getTimeStamps()
        fields = self._getFields()

        if not timeStamps is None:
            self._openTimeCollection()
            self._writeTimeStamps(timeStamps)
            for iTime, timeStamp in enumerate(timeStamps):
                self._openTimeGrid()
                self._writeGridTopology(cells)
                self._writeGridGeometry(vertices)
                for field in fields:
                    if field.vectorFieldType in ["Tensor6", "Matrix"]:
                        numComponents = field.data.shape[-1]
                        for iComponent in range(numComponents):
                            self._writeGridFieldComponent(
                                field, iTime, iComponent)
                    else:
                        self._writeGridField(field, iTime)
                self._closeTimeGrid()
            self._closeTimeCollection()
        else:
            iTime = None
            self._openTimeGrid()
            self._writeGridTopology(cells)
            self._writeGridGeometry(vertices)
            for field in fields:
                if field.vectorFieldType in ["Tensor6", "Matrix"]:
                    numComponents = field.data.shape[-1]
                    for iComponent in range(numComponents):
                        self._writeGridFieldComponent(field, iTime, iComponent)
                else:
                    self._writeGridField(field, iTime)
            self._closeTimeGrid()

        self._closeDomain()
        self._closeXdmf()
        self._close()
        return

    # PRIVATE METHODS ////////////////////////////////////////////////////

    def _close(self):
        self.h5.close()
        self.file.close()
        return

    def _getSpaceDim(self):
        vertices = self.h5["/geometry/vertices"]
        assert(2 == len(vertices.shape))
        return vertices.shape[1]

    def _getXdmfCellType(self, cells):
        """Get type of cell.
        """
        assert(2 == len(cells.shape))
        numCells, numCorners = cells.shape
        if "cell_dim" in cells.attrs:
            cellDim = cells.attrs["cell_dim"]
        else:
            # Use space dimension as a proxy for cell dimension.
            vertices = self.h5["/geometry/vertices"]
            assert(2 == len(vertices.shape))
            cellDim = vertices.shape[1]
        if 0 == cellDim and 1 == numCorners:
            cellType = "Polyvertex"
        elif 1 == cellDim and 2 == numCorners:
            cellType = "Polyline"
        elif 2 == cellDim and 3 == numCorners:
            cellType = "Triangle"
        elif 2 == cellDim and 4 == numCorners:
            cellType = "Quadrilateral"
        elif 3 == cellDim and 4 == numCorners:
            cellType = "Tetrahedron"
        elif 3 == cellDim and 8 == numCorners:
            cellType = "Hexahedron"
        else:
            cellType = "Unknown"
            print("WARNING: Unknown cell type with %d vertices and dimension %d." % (
                numCorners, cellDim))
        return cellType

    def _getXdmfVectorFieldType(self, vectorFieldString):
        """Get Xdmf vector field type.
        """
        vtype = "Matrix"
        if vectorFieldString.decode().lower() == "scalar":
            vtype = "Scalar"
        elif vectorFieldString.decode().lower() == "vector":
            vtype = "Vector"
        elif vectorFieldString.decode().lower() == "tensor":
            vtype = "Tensor6"
        return vtype

    def _getTimeStamps(self):
        """Get time stamps if they exist, otherwise return None.
        """
        timeStamps = None
        if "time" in self.h5:
            timeStamps = self.h5["time"][:]
        return timeStamps

    def _getFields(self):
        fields = []
        for group in ["vertex_fields", "cell_fields"]:
            if group in self.h5:
                vfields = self.h5[group]
                for name, dataset in vfields.items():
                    field = Field()
                    field.name = name
                    field.data = dataset[:]
                    field.domain = Field.groupToDomain[group]
                    if "vector_field_type" in dataset.attrs:
                        field.vectorFieldType = self._getXdmfVectorFieldType(
                            dataset.attrs["vector_field_type"])
                    else:
                        print(
                            "WARNING: Field '%s/%s' dataset missing 'vector_field_type' attribute. Guessing vector field type." % (group, name,))
                        field.vectorFieldType = "Matrix"
                        if len(dataset.shape) == 2 or len(dataset.shape) == 3:
                            numComponents = dataset.shape[-1]
                            if numComponents == 1:
                                field.vectorFieldType = "Scalar"
                            elif numComponents == 3:
                                field.vectorFieldType = "Vector"
                    fields.append(field)
        return fields

    def _openXdmf(self):
        """Write header and create Xdmf element.
        """
        import os

        filenameH5 = os.path.split(self.h5.filename)[-1]
        self.file.write(
            "<?xml version=\"1.0\" ?>\n"
            "<!DOCTYPE Xdmf SYSTEM \"Xdmf.dtd\" [\n"
            "<!ENTITY HeavyData \"%s\">\n"
            "]>\n"
            "\n"
            "<Xdmf>\n"
            % (filenameH5,)
        )
        return

    def _closeXdmf(self):
        """Close Xdmf element.
        """
        self.file.write(
            "</Xdmf>\n"
        )
        return

    def _openDomain(self, cells, vertices):
        self.file.write("  <Domain Name=\"domain\">\n")

        # Cells
        assert(2 == len(cells.shape))
        numCells, numCorners = cells.shape
        self.file.write(
            "    <DataItem Name=\"cells\" ItemType=\"Uniform\" Format=\"HDF\" NumberType=\"Float\" Precision=\"8\" Dimensions=\"%d %d\">\n"
            "      &HeavyData;:/topology/cells\n"
            "    </DataItem>\n"
            % (numCells, numCorners,)
        )

        # Vertices
        assert(2 == len(vertices.shape))
        numVertices, spaceDim = vertices.shape
        if 3 == spaceDim:
            self.file.write(
                "    <DataItem Name=\"vertices\" ItemType=\"Uniform\" Format=\"HDF\" Dimensions=\"%d %d\">\n"
                "      &HeavyData;:/geometry/vertices\n"
                "    </DataItem>\n"
                % (numVertices, spaceDim),
            )
        elif 2 == spaceDim:
            # Form vector with 3 components using x and y components
            # and then a fake z-component by multiplying the
            # x-component by zero.
            self.file.write(
                "    <DataItem Name=\"vertices\" ItemType=\"Function\" Dimensions=\"%d 3\" Function=\"JOIN($0, $1, $2)\">\n" % numVertices)
            # x component
            self.file.write(
                "      <DataItem Name=\"verticesX\" ItemType=\"Hyperslab\" Type=\"HyperSlab\" Dimensions=\"%d 1\">\n"
                "        <DataItem Dimensions=\"3 2\" Format=\"XML\">\n"
                "          0 0   1 1   %d 1\n"
                "        </DataItem>\n"
                "        <DataItem Dimensions=\"%d 1\" Format=\"HDF\">\n"
                "          &HeavyData;:/geometry/vertices\n"
                "        </DataItem>\n"
                "      </DataItem>\n"
                % (numVertices, numVertices, numVertices,)
            )

            # y component
            self.file.write(
                "      <DataItem Name=\"verticesY\" ItemType=\"Hyperslab\" Type=\"HyperSlab\" Dimensions=\"%d 1\">\n"
                "        <DataItem Dimensions=\"3 2\" Format=\"XML\">\n"
                "          0 1   1 1   %d 1\n"
                "        </DataItem>\n"
                "        <DataItem Dimensions=\"%d 1\" Format=\"HDF\">\n"
                "          &HeavyData;:/geometry/vertices\n"
                "        </DataItem>\n"
                "      </DataItem>\n"
                % (numVertices, numVertices, numVertices,))

            # z component
            self.file.write(
                "      <DataItem Name=\"verticesZ\" ItemType=\"Function\" Dimensions=\"%d 1\" Function=\"0*$0\">\n"
                "        <DataItem Reference=\"XML\">\n"
                "          /Xdmf/Domain/DataItem[@Name=\"vertices\"]/DataItem[@Name=\"verticesX\"]\n"
                "        </DataItem>\n"
                "      </DataItem>\n" % (numVertices,)
            )

            self.file.write("    </DataItem>\n")
        else:
            self._close()
            raise ValueError(
                "Unexpected spatial dimension %d when writing domain vertices." % spaceDim)
        return

    def _closeDomain(self):
        """Close domain element.
        """
        self.file.write(
            "  </Domain>\n"
        )
        return

    def _openTimeCollection(self):
        """Create Grid element for collection of time grids.
        """
        self.file.write(
            "    <Grid Name=\"TimeSeries\" GridType=\"Collection\" CollectionType=\"Temporal\">\n"
        )
        return

    def _closeTimeCollection(self):
        """Close Grid element for collection of time grids.
        """
        self.file.write(
            "    </Grid>\n"
        )
        return

    def _writeTimeStamps(self, tstamps):
        """Write time stamps.
        """
        self.file.write(
            "      <Time TimeType=\"List\">\n"
            "        <DataItem Format=\"XML\" NumberType=\"Float\" Dimensions=\"%d\">\n"
            "        "
            % (tstamps.shape[0],)
        )
        for t in tstamps:
            self.file.write("  %16.8e" % t)
        self.file.write(
            "\n"
            "        </DataItem>\n"
            "      </Time>\n"
        )
        return

    def _openTimeGrid(self):
        """Create Grid element for a single time step.
        """
        self.file.write(
            "      <Grid Name=\"domain\" GridType=\"Uniform\">\n"
        )
        return

    def _closeTimeGrid(self):
        """Close Grid element for a single time step.
        """
        self.file.write(
            "      </Grid>\n"
        )
        return

    def _writeGridTopology(self, cells):
        """Write topology information for current grid.
        """
        cellType = self._getXdmfCellType(cells)
        assert(2 == len(cells.shape))
        numCells = cells.shape[0]

        self.file.write(
            "        <Topology TopologyType=\"%s\" NumberOfElements=\"%d\">\n"
            "          <DataItem Reference=\"XML\">\n"
            "            /Xdmf/Domain/DataItem[@Name=\"cells\"]\n"
            "          </DataItem>\n"
            "        </Topology>\n"
            % (cellType, numCells,)
        )
        return

    def _writeGridGeometry(self, vertices):
        """Write vertices information for current grid.
        """
        self.file.write(
            "        <Geometry GeometryType=\"XYZ\">\n"
            "          <DataItem Reference=\"XML\">\n"
            "            /Xdmf/Domain/DataItem[@Name=\"vertices\"]\n"
            "          </DataItem>\n"
            "        </Geometry>\n"
        )
        return

    def _writeGridFieldComponent(self, field, iTime, iComponent):
        """Write single component of field for current time step.
        """
        if field.vectorFieldType == "Vector":
            components = ["_x", "_y", "_z"]
            componentName = field.name + components[iComponent]
        elif field.vectorFieldType == "Tensor6":
            spaceDim = self._getSpaceDim()
            if spaceDim == 2:
                components = ["_xx", "_yy", "_xy"]
            elif spaceDim == 3:
                components = ["_xx", "_yy", "_zz", "_xy", "_yz", "_xz"]
            else:
                self._close()
                raise ValueError(
                    "Unexpected spatial dimension %d for field component names." % spaceDim)
            componentName = field.name + components[iComponent]
        elif field.vectorFieldType == "Matrix":
            componentName = field.name + "_%d" % iComponent
        else:
            self._close()
            raise ValueError(
                "Unexpected vector field type '%s' for field component names." % field.vectorFieldType)
        h5Name = "/" + Field.domainToGroup[field.domain] + "/" + field.name
        if iTime is None:
            assert(2 == len(field.data.shape))
            numPoints, numComponents = field.data.shape
            numTimeSteps = 1
        else:
            assert(3 == len(field.data.shape))
            numTimeSteps, numPoints, numComponents = field.data.shape

        self.file.write(
            "        <Attribute Name=\"%(componentName)s\" Type=\"Scalar\" Center=\"%(domain)s\">\n"
            "          <DataItem ItemType=\"HyperSlab\" Dimensions=\"1 %(numPoints)d 1\" Type=\"HyperSlab\">\n"
            "            <DataItem Dimensions=\"3 3\" Format=\"XML\">\n"
            "              %(iTime)d 0 %(iComponent)d    1 1 1    1 %(numPoints)d 1\n"
            "            </DataItem>\n"
            "            <DataItem DataType=\"Float\" Precision=\"8\" Dimensions=\"%(numTimeSteps)d %(numPoints)d %(numComponents)d\" Format=\"HDF\">\n"
            "              &HeavyData;:%(h5Name)s\n"
            "            </DataItem>\n"
            "          </DataItem>\n"
            "        </Attribute>\n"
            % {"componentName": componentName,
               "domain": field.domain,
               "numPoints": numPoints,
               "iTime": iTime,
               "iComponent": iComponent,
               "numTimeSteps": numTimeSteps,
               "numComponents": numComponents,
               "h5Name": h5Name,
               }
        )
        return

    def _writeGridField(self, field, iTime):
        """Write field for current time step.
        """
        gridRef = "/Xdmf/Domain/Grid" if iTime is None else "/Xdmf/Domain/Grid/Grid[1]"
        self.file.write(
            "        <Attribute Name=\"%s\" Type=\"%s\" Center=\"%s\">\n"
            % (field.name, field.vectorFieldType, field.domain,)
        )
        h5Name = "/" + Field.domainToGroup[field.domain] + "/" + field.name
        iStep = iTime
        if iTime is None:
            iStep = 0
            if 2 == len(field.data.shape):
                numPoints, numComponents = field.data.shape
                numTimeSteps = 1
            elif 3 == len(field.data.shape):
                numTimeSteps, numPoints, numComponents = field.data.shape
            else:
                raise ValueError(
                    "Unexpected shape for dataset '%s'." % field.name)
        else:
            assert(3 == len(field.data.shape))
            numTimeSteps, numPoints, numComponents = field.data.shape

        if 2 == self._getSpaceDim() and field.vectorFieldType == "Vector":

            self.file.write(
                "          <DataItem ItemType=\"Function\" Dimensions=\"%d 3\" Function=\"JOIN($0, $1, $2)\">\n"
                % (numPoints,)
            )
            # x component
            self.file.write(
                "            <DataItem ItemType=\"HyperSlab\" Dimensions=\"%(numPoints)d 1\" Type=\"HyperSlab\">\n"
                "              <DataItem Dimensions=\"3 3\" Format=\"XML\">\n"
                "                %(iStep)d 0 0    1 1 1    1 %(numPoints)d 1\n"
                "              </DataItem>\n"
                "              <DataItem DataType=\"Float\" Precision=\"8\" Dimensions=\"%(numTimeSteps)d %(numPoints)d %(numComponents)d\" Format=\"HDF\">\n"
                "                &HeavyData;:%(h5Name)s\n"
                "              </DataItem>\n"
                "            </DataItem>\n"
                % {"numTimeSteps": numTimeSteps, "numPoints": numPoints, "iStep": iStep, "numComponents": numComponents, "h5Name": h5Name}
            )

            # y component
            self.file.write(
                "            <DataItem ItemType=\"HyperSlab\" Dimensions=\"%(numPoints)d 1\" Type=\"HyperSlab\">\n"
                "              <DataItem Dimensions=\"3 3\" Format=\"XML\">\n"
                "                %(iStep)d 0 1    1 1 1    1 %(numPoints)d 1\n"
                "              </DataItem>\n"
                "              <DataItem DataType=\"Float\" Precision=\"8\" Dimensions=\"%(numTimeSteps)d %(numPoints)d %(numComponents)d\" Format=\"HDF\">\n"
                "                &HeavyData;:%(h5Name)s\n"
                "              </DataItem>\n"
                "            </DataItem>\n"
                % {"numTimeSteps": numTimeSteps, "numPoints": numPoints, "iStep": iStep, "numComponents": numComponents, "h5Name": h5Name}
            )

            # z component
            self.file.write(
                "            <DataItem ItemType=\"Function\" Dimensions=\"%(numPoints)d 1\" Function=\"0*$0\">\n"
                "              <DataItem Reference=\"XML\">\n"
                "                %(gridRef)s/Attribute[@Name=\"%(name)s\"]/DataItem[1]/DataItem[1]\n"
                "              </DataItem>\n"
                "            </DataItem>\n"
                % {"numPoints": numPoints, "name": field.name, "gridRef": gridRef}
            )

            # close
            self.file.write(
                "          </DataItem>\n"
                "        </Attribute>\n"
            )

        else:
            self.file.write(

                "          <DataItem ItemType=\"HyperSlab\" Dimensions=\"1 %(numPoints)d %(numComponents)d\" Type=\"HyperSlab\">\n"
                "            <DataItem Dimensions=\"3 3\" Format=\"XML\">\n"
                "              %(iStep)d 0 0    1 1 1    1 %(numPoints)d %(numComponents)d\n"
                "            </DataItem>\n"
                "            <DataItem DataType=\"Float\" Precision=\"8\" Dimensions=\"%(numTimeSteps)d %(numPoints)d %(numComponents)d\" Format=\"HDF\">\n"
                "              &HeavyData;:%(h5Name)s\n"
                "            </DataItem>\n"
                "          </DataItem>\n"
                "        </Attribute>\n"
                % {"numTimeSteps": numTimeSteps, "numPoints": numPoints, "iStep": iStep, "numComponents": numComponents, "h5Name": h5Name}
            )

            return


# End of file
